server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java

```diff
@@ -31,18 +31,34 @@
 import java.util.Objects;
 
 /**
- * {@link ValuesSourceRegistry} holds the mapping from {@link ValuesSourceType}s to {@link AggregatorSupplier}s.  DO NOT directly
+ * {@link ValuesSourceRegistry} holds the mapping from {@link ValuesSourceType}s to {@link ComponentSupplier}s.  DO NOT directly
  * instantiate this class, instead get an already-configured copy from {@link QueryShardContext#getValuesSourceRegistry()}, or (in the case
  * of some test scenarios only) directly from {@link SearchModule#getValuesSourceRegistry()}
  *
  */
 public class ValuesSourceRegistry {
 
-    public interface CompositeSupplier {
+    /**
+     * {@link ComponentSupplier} serves as a marker for what the {@link ValuesSourceRegistry} holds to construct aggregator instances.
+     * The aggregators for each aggregation should all share a signature, and that signature should be used to create an ComponentSupplier
+     * for that aggregation.  Alternatively, if an existing supplier has a matching signature, please re-use that.
+     *
+     * In many cases, this can be a simple wrapper over the aggregator constructor.  If that is sufficient, please just use a reference to
+     * the constructor.  Implementing a supplier (typically as a lambda) should only be necessary when factors besides the
+     * {@link ValuesSourceType} are necessary for selecting the correct aggregator implementation.  This happens in terms for example where
+     * we make decisions based on the availability of global ordinals.
+     *
+     * The suppliers are responsible for any casting of {@link ValuesSource} that needs to happen.  They must accept a base
+     * {@link ValuesSource} * instance.  The suppliers may perform additional logic to configure the aggregator as needed, such as in
+     * {@link org.elasticsearch.search.aggregations.bucket.terms.TermsAggregatorFactory} deciding the execution mode.
+     *
+     * There is ongoing work to  normalize aggregator constructor signatures, and thus reduce the number of ComponentSupplier interfaces.
+     */
+    public interface ComponentSupplier {
```

Comment:
Do you think we need the marker interface? Like, maybe the `<T>` below can just be `T`?